libraries/Parsers/Microsoft.Bot.Builder.Parsers.LU/Parser/LuParser.cs

```diff
@@ -0,0 +1,448 @@
+            var parser = new LUFileParser(tokens);
+            parser.BuildParseTree = true;
+            return parser.file();
+        }
+
+        private static void ExtractSectionBody(List<Section> sections, string content)
+        {
+            var originList = Regex.Split(content, @"\r?\n");
+            var qnaSectionIndex = 0;
+            foreach (var sectionIter in sections.Select((x, i) => new { Value = x, Index = i }))
+            {
+                var section = sectionIter.Value;
+                if (section is SimpleIntentSection || section is NestedIntentSection || section is QnaSection)
+                {
+                    var startLine = section.Range.Start.Line - 1;
+                    int stopLine = -1;
+                    if (sectionIter.Index + 1 < sections.Count)
+                    {
+                        stopLine = sections[sectionIter.Index + 1].Range.Start.Line - 1;
+                        if (startLine < 0 || startLine > stopLine)
+                        {
+                            throw new IndexOutOfRangeException();
+                        }
+                    }
+                    else
+                    {
+                        stopLine = originList.Length;
+                    }
+
+                    section.Range.End.Line = stopLine;
+                    section.Range.End.Character = originList[stopLine - 1].Length;
+
+                    var destList = new List<string>();
+                    if (section is QnaSection)
+                    {
+                        destList = originList.Skip(startLine).Take(stopLine - startLine).ToList();
+
+                        // TODO: maybe change the model so Id is int
+#pragma warning disable CA1305 // Specify IFormatProvider
+                        section.Id = qnaSectionIndex.ToString();
+#pragma warning restore CA1305 // Specify IFormatProvider
+                        qnaSectionIndex++;
+                    }
+                    else
+                    {
+                        destList = originList.Skip(startLine + 1).Take(stopLine - (startLine + 1)).ToList();
+                    }
+
+                    section.Body = string.Join(Environment.NewLine, destList);
+
+                    if (section is NestedIntentSection nestedIntentSection)
+                    {
+                        // TODO: check if this downcast can re-cast
+                        var sectionList = nestedIntentSection.SimpleIntentSections.ToList<Section>();
+                        ExtractSectionBody(sectionList, string.Join(Environment.NewLine, originList.Skip(0).Take(stopLine)));
+                    }
+                }
+            }
+        }
+
+        private static bool IsSectionEnabled(List<Section> sections)
+        {
+            var modelInfoSections = sections.Where(s => s.SectionType == SectionType.ModelInfoSection);
+            bool enableSections = false;
+
+            if (modelInfoSections.Any())
```

Comment:
To avoid multiple enumeration, consider converting modelInfoSections to list or array.